experiments/esp32_mp/src/main.c

```diff
@@ -0,0 +1,131 @@
+#include <zephyr.h>
+#include <misc/printk.h>
+
+#define _REG(base, off) (*(volatile u32_t *)((base) + (off)))
+
+#define RTC_CNTL_BASE             0x3ff48000
+#define RTC_CNTL_OPTIONS0     _REG(RTC_CNTL_BASE, 0x0)
+#define RTC_CNTL_SW_CPU_STALL _REG(RTC_CNTL_BASE, 0xac)
+
+#define DPORT_BASE                 0x3ff00000
+#define DPORT_APPCPU_CTRL_A    _REG(DPORT_BASE, 0x02C)
+#define DPORT_APPCPU_CTRL_B    _REG(DPORT_BASE, 0x030)
+#define DPORT_APPCPU_CTRL_C    _REG(DPORT_BASE, 0x034)
+
+/* Two fields with same naming conventions live in two different
+ * registers and have different widths...
+ */
+#define RTC_CNTL_SW_STALL_APPCPU_C0  (0x03 << 0)  /* reg: OPTIONS0 */
+#define RTC_CNTL_SW_STALL_APPCPU_C1  (0x3f << 20) /* reg: SW_CPU_STALL */
+
+#define DPORT_APPCPU_CLKGATE_EN  BIT(0)
+#define DPORT_APPCPU_RUNSTALL    BIT(0)
+#define DPORT_APPCPU_RESETTING   BIT(0)
+
+
+/* These calls are ROM-resident and have fixed addresses. No, I don't
```

Comment:
They usually copy the code to esp-idf and patch it there.  Of course, if you're using the ROM functions yourself and they end up being buggy, you're out of luck.

(Also, yeah, I'm adding a `esp32_rom_` prefix to ROM functions now.  Changed that inside the linker script, and the prototype for those functions are in the SOC directory, instead of esp-idf.)